actors/evm/src/interpreter/precompiles.rs

```diff
@@ -336,16 +356,53 @@ fn read_right_pad<'a>(input: impl Into<Cow<'a, [u8]>>, len: usize) -> Cow<'a, [u
 
 // --- Precompiles ---
 
-/// Read right padded BE encoded low u64 ID address from a u256 word
-/// returns encoded CID or an empty array if actor not found
-fn get_actor_code_cid<RT: Runtime>(
+/// Read right padded BE encoded low u64 ID address from a u256 word.
+/// Returns variant of [`BuiltinType`] encoded as a u256 word.
+fn get_actor_type<RT: Runtime>(
     system: &mut System<RT>,
     input: &[u8],
     _: PrecompileContext,
 ) -> PrecompileResult {
+    const LAST_SYSTEM_ACTOR_ID: u64 = 32;
+
     let id_bytes: [u8; 32] = read_right_pad(input, 32).as_ref().try_into().unwrap();
     let id = Parameter::<u64>::try_from(&id_bytes)?.0;
-    Ok(system.rt.get_actor_code_cid(&id).unwrap_or_default().to_bytes())
+
+    if id < LAST_SYSTEM_ACTOR_ID {
+        // known to be system actors
+        Ok(NativeType::System.word_vec())
+    } else {
+        // resolve type from code CID
+        let builtin_type = system
+            .rt
+            .get_actor_code_cid(&id)
+            .and_then(|cid| system.rt.resolve_builtin_actor_type(&cid));
+
+        let builtin_type = match builtin_type {
+            Some(t) => match t {
+                Type::Account => NativeType::Account,
+                Type::System => NativeType::System,
+                Type::Embryo => NativeType::Embryo,
+                Type::EVM => NativeType::EVMContract,
+                Type::Miner => NativeType::StorageProvider,
+                // Others
+                Type::PaymentChannel | Type::Multisig => NativeType::OtherTypes,
+                // Singletons
+                Type::Market
+                | Type::Power
+                | Type::Init
+                | Type::Cron
+                | Type::Reward
+                | Type::VerifiedRegistry
+                | Type::DataCap => NativeType::System,
+                // REMOVEME this might be useful to differentiate? though we already have opcodes that do all the calls we need to do for this
```

Comment:
IMO, we shouldn't differentiate this. I can't think of a reason an EVM actor would care.

Reply:
There are a few small cases where it might be nice, but could happily be done without being able the differentiate. (plus this breaks the pattern of "singletons are system".